### ISSUE TITLE
Bug: `no-redundant-type-constituents` incorrectly marks `& string` as redundant

### ISSUE THREAD

**Issue by worstpractice** [typescript-eslint/typescript-eslint#7580]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.ts&code=LAKApgHgDg9gTgFwAQIJ5TEgggGxzAdwGcA5AVwFsAjMOAaTFSKQF4kBvUJJOMAQwAmMAHY5USANoBjIkQAKcGBkSoSfCpgCWwpAGtGMAGZIAwgGULaHGAAiYKTj5w%2BCTSIC6AfgBcpi2atbe0dnVxFpWQUlWjQ1DXcAblAAXySQUEhYRBR0TBtNIj48QlJKGnpGZjZOEG4AejqkABUACwKkdr4kKkFmswBRIhxtZFpFOCRAXg3AUp2uJAa5hqQASQQAcmYEFsxDIqJMWCJNVwA3TAAKXkFfKjIAcw6dLcxh4WQ4MmsASgAaRcaBGAKCIiAhQtoHi1CChtkhBAJjm4dEYkAADABkSFBcAhqI6zAAVMIYAgCTwwAIyMIBHw3gA6OZXISicQReSKZSxdRaHT6VAo8yWVDWOwOJwuJFITHYiFeXyCgLCoJi8HhGTs6IqOJgRIpUAZaDwZBSEHIFp8IhYcjUWgMcTVOYABl8a0mgAkdmHtPn4uFIYRlWhrP4gVL68CG7Im4SgpAETSAu1EXy4fDEa3lROsJDmy3p22MNKgJatdqdJB3GAwAR9JBjeBIQCg5EW6s3TCD47QIbGwGszrXoMMpMcxH6nIoCBSkEICE9YXzW2gMEgUaj8oVimmAxUmHiEDBlyy4VIpGAoMgZcI7kQGelw1ljabR3Bx4nfGuiqnSjbt1VsxarVudppEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNYmvAIYtaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=&tokens=false

### Repro Code

```typescript
export type AllowsNumberKeys = {
  readonly [cssPropertyName in keyof CSSStyleDeclaration]?: CSSStyleDeclaration[cssPropertyName];
};

export type DisallowsNumberKeys = {
  // This is a bad TSEslint error 💥
  //
  // It's the false positive (read: bug in the lint rule),
  // demonstrating how the addition of `& string` is *not* redundant.
  readonly [cssPropertyName in keyof CSSStyleDeclaration & string]?: CSSStyleDeclaration[cssPropertyName];
}

export const hasANumberKey = {
  0: '👈 this key is a number',
};

export const wideKeys: AllowsNumberKeys = hasANumberKey;

// This is a good TS error ✅
//
// Considering we've explicitly narrowed down the key
// type of `DisallowsNumberKeys` to only accept strings.
export const narrowKeys: DisallowsNumberKeys = hasANumberKey;
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-redundant-type-constituents": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {}
}
```

### Expected Result

No lint error, considering that adding `& string` to the key type is *not* redundant.

### Actual Result

A false positive lint error claiming that the addition of `& string` is redundant, which it is *not*.

### Additional Info

Hello 👋

I searched and found no apparent overlap with any previous issue. Apologies if I missed anything and this turns out to be a duplicate after all.

**Comment by bradzacher** [comment]
In this case you should use `Extract<keyof CSSStyleDeclaration, string>` as it's infinitely clearer what your intent is - to extract the members of the union that are assignable to `string`.

Honestly I'm okay with this erroring because the code can be made much clearer with a minor change.

Cc @JoshuaKGoldberg

**Comment by JoshuaKGoldberg** [comment]
Yeah +1, this is a pretty rare case in part because IME most folks writing code like it will use `Extract`. Is there a snippet demonstrating the bug that's simpler?

**Comment by bradzacher** [comment]
[playground](https://typescript-eslint.io/play/#ts=5.2.2&fileType=.ts&code=C4TwDgpgBAKlC8UDkBDJUA%2BUCMBuAsAFCiRQCqCsUAZFAM7ABOAlgHYDmBhRAxgPasGUAGZ8%2BALnKU8UAPSyoEAB6QewCABNFjRn0ZEgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNYmvAIYtaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=&tokens=false)

when you intersect with a union - `(A | B) & C` - for the resulting type TS will extract all types from the union that are assignable to `C`.

So yeah the rule technically shouldn't report on this case because the intersection does something - we could detect if all the union members are assignable - but it's not going to be fast or simple.
